drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,713 @@
+
+static const struct i2c_driver_api i2c_esp32_driver_api = {
+	.configure = i2c_esp32_configure,
+	.transfer = i2c_esp32_transfer,
+};
+
+#ifdef CONFIG_I2C_0
+DEVICE_DECLARE(i2c_esp32_0);
+
+static void i2c_esp32_connect_irq_0(void)
+{
+	IRQ_CONNECT(CONFIG_I2C_ESP32_0_IRQ, 1, i2c_esp32_isr,
+		    DEVICE_GET(i2c_esp32_0), 0);
+}
+
+static const struct i2c_esp32_config i2c_esp32_config_0 = {
+	.index = 0,
+	.connect_irq = i2c_esp32_connect_irq_0,
+	.sig = {
+		.sda_out = I2CEXT0_SDA_OUT_IDX,
+		.sda_in = I2CEXT0_SDA_IN_IDX,
+		.scl_out = I2CEXT0_SCL_OUT_IDX,
+		.scl_in = I2CEXT0_SCL_IN_IDX,
+	},
+	.pins = {
+		.scl = CONFIG_I2C_ESP32_0_SCL_PIN,
+		.sda = CONFIG_I2C_ESP32_0_SDA_PIN,
+	},
+	.enable_bits = {
+		.clk = DPORT_I2C_EXT0_CLK_EN,
+		.rst = DPORT_I2C_EXT0_RST,
+	},
+	.mode = {
+		.tx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_0_TX_LSB_FIRST),
+		.rx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_0_RX_LSB_FIRST),
+	},
+	.irq = {
+		.source = ETS_I2C_EXT0_INTR_SOURCE,
+		.line = CONFIG_I2C_ESP32_0_IRQ,
+	},
+	.default_config = CONFIG_I2C_0_DEFAULT_CFG,
+};
+
+static struct i2c_esp32_data i2c_esp32_data_0;
+
+DEVICE_AND_API_INIT(i2c_esp32_0, CONFIG_I2C_0_NAME, &i2c_esp32_init,
+		    &i2c_esp32_data_0, &i2c_esp32_config_0,
+		    POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
+		    &i2c_esp32_driver_api);
+#endif /* CONFIG_I2C_0 */
+
+#ifdef CONFIG_I2C_1
+DEVICE_DECLARE(i2c_esp32_1);
+
+static void i2c_esp32_connect_irq_1(void)
+{
+	IRQ_CONNECT(CONFIG_I2C_ESP32_0_IRQ, 1, i2c_esp32_isr,
+		    DEVICE_GET(i2c_esp32_1), 0);
+}
+
+static const struct i2c_esp32_config i2c_esp32_config_1 = {
+	.index = 1,
+	.connect_irq = i2c_esp32_connect_irq_1,
+	.sig = {
+		.sda_out = I2CEXT1_SDA_OUT_IDX,
+		.sda_in = I2CEXT1_SDA_IN_IDX,
+		.scl_out = I2CEXT1_SCL_OUT_IDX,
+		.scl_in = I2CEXT1_SCL_IN_IDX,
+	},
+	.pins = {
+		.scl = CONFIG_I2C_ESP32_1_SCL_PIN,
+		.sda = CONFIG_I2C_ESP32_1_SDA_PIN,
+	},
+	.enable_bits = {
+		.clk = DPORT_I2C_EXT1_CLK_EN,
+		.rst = DPORT_I2C_EXT1_RST,
+	},
+	.mode = {
+		.tx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_1_TX_LSB_FIRST),
+		.rx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_1_RX_LSB_FIRST),
+	},
+	.irq = {
+		.source = ETS_I2C_EXT1_INTR_SOURCE,
+		.line = CONFIG_I2C_ESP32_1_IRQ,
+	},
+	.default_config = CONFIG_I2C_1_DEFAULT_CFG,
+};
+
+static struct i2c_esp32_data i2c_esp32_data_1;
+
+DEVICE_AND_API_INIT(i2c_esp32_1, CONFIG_I2C_1_NAME, &i2c_esp32_init,
+		    &i2c_esp32_data_1, &i2c_esp32_config_1,
+		    POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
+		    &i2c_esp32_driver_api);
+#endif /* CONFIG_I2C_1 */
+
+static int i2c_esp32_init(struct device *dev)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	int key = irq_lock();
+
+	k_sem_init(&data->fifo_sem, 1, 1);
+	k_sem_init(&data->transfer_sem, 1, 1);
+
+	irq_disable(config->irq.line);
+
+	/* Even if irq_enable() is called on config->irq.line, disable
+	 * interrupt sources in the I2C controller.
+	 */
```

Comment:
Is there a specific esp32 behavior that requires the driver to specifically disable the irq to re-enable them at configure time?
Usually, irq_enable() is ran after IRQ_CONNECT() and disabling the interrupts from the controller is sufficient to silence it, thus why I am asking.

Reply:
ESP32 has two CPUs, and both can fire interrupts.  While the second CPU is currently halted, there are plans to support SMP in the near future.